examples/chatgpt/rag-quickstart/azure/Azure_AI_Search_with_Azure_Functions_and_GPT_Actions_in_ChatGPT.ipynb

```diff
@@ -0,0 +1,1330 @@
+            "cell_type": "markdown",
+            "metadata": {},
+            "source": [
+                "Finally, we can write a function that safely handles embedding requests, even when the input text is longer than the maximum context length, by chunking the input tokens and embedding each chunk individually. The average flag can be set to True to return the weighted average of the chunk embeddings, or False to simply return the unmodified list of chunk embeddings.\n",
+                "\n",
+                "> Note: there are other more sophisticated techniques you can take here, including:\n",
+                "> - using GPT-4o to capture images/chart descriptions for embedding.\n",
+                "> - keeping text overlap between the chunks to minimize cutting off important context.\n",
+                "> - chunking based on paragraphs or sections.\n",
+                "> - adding more descriptive metadata about each article."
+            ]
+        },
+        {
+            "cell_type": "code",
+            "execution_count": null,
+            "metadata": {},
+            "outputs": [],
+            "source": [
+                "EMBEDDING_CTX_LENGTH = 8191\n",
+                "EMBEDDING_ENCODING='cl100k_base'"
+            ]
+        },
+        {
+            "cell_type": "code",
+            "execution_count": null,
+            "metadata": {},
+            "outputs": [],
+            "source": [
+                "def generate_embeddings(text, model):\n",
+                "    # Generate embeddings for the provided text using the specified model\n",
+                "    embeddings_response = openai_client.embeddings.create(model=model, input=text)\n",
+                "    # Extract the embedding data from the response\n",
+                "    embedding = embeddings_response.data[0].embedding\n",
+                "\n",
+                "    return embedding\n",
+                "def len_safe_get_embedding(text, model=embeddings_model, max_tokens=EMBEDDING_CTX_LENGTH, encoding_name=EMBEDDING_ENCODING):\n",
+                "    # Initialize lists to store embeddings and corresponding text chunks\n",
+                "    chunk_embeddings = []\n",
+                "    chunk_texts = []\n",
+                "    \n",
+                "    # Iterate over chunks of tokens from the input text\n",
+                "    for chunk in chunked_tokens(text, chunk_length=max_tokens, encoding_name=encoding_name):\n",
+                "        # Generate embeddings for each chunk and append to the list\n",
+                "        chunk_embeddings.append(generate_embeddings(chunk, model=model))\n",
+                "        # Decode the chunk back to text and append to the list\n",
+                "        chunk_texts.append(tiktoken.get_encoding(encoding_name).decode(chunk))\n",
+                "    # Return the list of chunk embeddings and the corresponding text chunks\n",
+                "    return chunk_embeddings, chunk_texts\n"
```

Comment:
Improve readability. most of the text here is comments, which makes it hard to digest.